diff -Nur -X dontdiff a/drivers/fingerprint/vfs61xx.c b/drivers/fingerprint/vfs61xx.c
--- a/drivers/fingerprint/vfs61xx.c	2016-06-01 21:15:08.562718000 -0400
+++ b/drivers/fingerprint/vfs61xx.c	2016-06-03 18:47:19.887779781 -0400
@@ -296,7 +296,7 @@
 }
 
 /* Return no.of bytes read >0. negative integer incase of error. */
-inline ssize_t vfsspi_readSync(struct vfsspi_devData *vfsSpiDev,
+static inline ssize_t vfsspi_readSync(struct vfsspi_devData *vfsSpiDev,
 	unsigned char *buf, size_t len)
 {
 	int status = 0;
diff -Nur -X dontdiff a/drivers/tty/serial/msm_serial_hs.c b/drivers/tty/serial/msm_serial_hs.c
--- a/drivers/tty/serial/msm_serial_hs.c	2016-06-03 13:47:25.501153000 -0400
+++ b/drivers/tty/serial/msm_serial_hs.c	2016-06-03 18:48:41.969339392 -0400
@@ -106,7 +106,7 @@
 char *g_szSysTime;
 char *g_szRefTime;	
 
-inline void UpdateTime(char *pchBuffer, int iLen)
+static inline void UpdateTime(char *pchBuffer, int iLen)
 {
 	struct timespec tSysTimespec;
 	char *pEnv[3];
diff -Nur -X dontdiff a/drivers/video/msm/mdss/mdss_mdp_pp.c b/drivers/video/msm/mdss/mdss_mdp_pp.c
--- a/drivers/video/msm/mdss/mdss_mdp_pp.c	2016-06-01 21:15:08.687721000 -0400
+++ b/drivers/video/msm/mdss/mdss_mdp_pp.c	2016-06-03 18:49:39.162426464 -0400
@@ -469,7 +469,7 @@
 
 static u32 last_sts, last_state;
 
-inline int linear_map(int in, int *out, int in_max, int out_max)
+static inline int linear_map(int in, int *out, int in_max, int out_max)
 {
 	if (in < 0 || !out || in_max <= 0 || out_max <= 0)
 		return -EINVAL;
diff -Nur -X dontdiff a/scripts/fips_crypto_utils.c b/scripts/fips_crypto_utils.c
--- a/scripts/fips_crypto_utils.c	2016-06-01 21:14:32.973997000 -0400
+++ b/scripts/fips_crypto_utils.c	2016-06-03 18:54:11.663609041 -0400
@@ -13,6 +13,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
 
 int main (int argc, char **argv)
