From 4b1522d0745fbb37a6c49cae61b7c5830c04d18f Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 10 Feb 2015 08:02:25 -0500
Subject: [PATCH 09/58] define HAVE_GETAUXVAL for getentropy_fallback

---
 libc/bionic/getentropy_linux.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/libc/bionic/getentropy_linux.c b/libc/bionic/getentropy_linux.c
index 1fb0c1d..78f0653 100644
--- a/libc/bionic/getentropy_linux.c
+++ b/libc/bionic/getentropy_linux.c
@@ -22,6 +22,7 @@
 
 #define	_POSIX_C_SOURCE	199309L
 #define	_GNU_SOURCE	1
+#define	HAVE_GETAUXVAL
 #include <sys/cdefs.h>
 #include <sys/types.h>
 #include <sys/param.h>
-- 
2.10.0

